### PR TITLE
[BUG] Fix single trace zero division error

### LINKED ISSUE #252: [BUG] Index-based colormodes raise `ZeroDivisionError` for single-trace plots
All index-based colormodes raise a `ZeroDivisionError` when plotting a ridgeline plot with either a single trace or a single trace per row.

The index-based colormodes are:

```python
colormodes = [
    "row-index",
    "trace-index",
    "trace-index-row-wise",
]
```

MVE:

```pycon
>>> from ridgeplot import ridgeplot
>>> ridgeplot(samples=[[1, 2, 2, 3]], colormode=colormodes[0])
Traceback (most recent call last):
...
  File "[...]/ridgeplot/_color/interpolation.py", line 68, in _interpolate_row_index
    [((ctx.n_rows - 1) - ith_row) / (ctx.n_rows - 1)] * len(row)
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
ZeroDivisionError: division by zero
```

### PR DESCRIPTION
**Pull request by sstephanyy**
## Description

To prevent the `ZeroDivisionError` from occurring when using index-based colormodes with a **single trace** or a **single row of data,** the code has been modified to **check the number of rows before performing any calculations**. If only one row is present, a default color value (like 0.0) is returned instead of attempting to divide by zero.

## Related issues

Closes #252 

## PR check list

- [X] Read the [contributing guidelines](https://ridgeplot.readthedocs.io/en/latest/development/contributing.html).
  - You don't have to read the whole thing, but it's a good idea to skim it. And definitely take a look at it if you're experiencing any issues getting your local environment up and running.
- [X] Provided the relevant details in the PR's description.
- [X] Referenced relevant issues in the PR's description.
- [X] Added tests for all my changes.
  - The CI will fail unless 100% of all new code is covered by the tests.
- [X] Updated the docs for relevant changes.
  - [X] New modules (or renamed ones) are included in `docs/api/internal/`.
  - [X] New public functions/classes/variables are documented in `docs/api/index.rst`.
  - [X] Added the appropriate `versionadded`, `versionchanged`, or `deprecated` [directives](http://www.sphinx-doc.org/en/stable/markup/para.html#directive-versionadded) to docstrings.
    - The version should be the next release version, which you can infer by bumping the minor version in `MAJOR.MINOR.PATCH` (e.g., if the current version is `0.2.3`, the next release will be `0.3.0`).
- [X] Changes are documented in `docs/reference/changelog.md`.
  - Please try to follow the conventions laid out in the file. In doubt, just ask!
- [X] Consider granting [push permissions to your PR's branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork), so maintainers can help you out if needed.
- [X] The CI check are all passing, or I'm working on fixing them!
- [X] I have reviewed my own code! 🤠

----
📚 Documentation preview 📚: https://ridgeplot--268.org.readthedocs.build/en/268/

### CODE REVIEW

**Comment by tpvasconcelos** [comment]
Hey @sstephanyy thanks a lot for your pull request!

I see that some CI checks are still failing. Let me know if you need me to look into it 👍

**Comment by sstephanyy** [comment]
> I see that some CI checks are still failing. Let me know if you need me to look into it 👍

@tpvasconcelos, I solved a mypy CI failing check, but 1 CI failure remains :/

**Comment by tpvasconcelos** [comment]
@sstephanyy just noticed that you added your changes to the wrong section of the changelog. No worries of course but this would mean that your changes would not be listed in the next release notes 😨 

Fixed in bba6be22c124d47c11f95fcdb1c602a01325369d ✅

**Comment by tpvasconcelos** [comment]
@sstephanyy I simplified your test cases a bit by using `pytest.mark.parametrize` (take a look).

File: src/ridgeplot/_color/interpolation.py [hunk 1]
```diff
@@ -115,13 +115,17 @@ def _mul(a: tuple[Numeric, ...], b: tuple[Numeric, ...]) -> tuple[Numeric, ...]:
 
 
 def _interpolate_row_index(ctx: InterpolationContext) -> ColorscaleInterpolants:
+    if ctx.n_rows <= 1:
```
Comment (tpvasconcelos) [inline src/ridgeplot/_color/interpolation.py:118]:
```suggestion
    if ctx.n_rows == 1:
```

File: src/ridgeplot/_color/interpolation.py [hunk 2]
```diff
@@ -115,13 +115,17 @@ def _mul(a: tuple[Numeric, ...], b: tuple[Numeric, ...]) -> tuple[Numeric, ...]:
...
+        return [[0.0] * len(row) for row in ctx.densities]
     return [
         [((ctx.n_rows - 1) - ith_row) / (ctx.n_rows - 1)] * len(row)
         for ith_row, row in enumerate(ctx.densities)
     ]
 
 
 def _interpolate_trace_index(ctx: InterpolationContext) -> ColorscaleInterpolants:
+    if ctx.n_traces <= 1:
```
Comment (tpvasconcelos) [inline src/ridgeplot/_color/interpolation.py:127]:
```suggestion
    if ctx.n_traces == 1:
```

File: src/ridgeplot/_color/interpolation.py [hunk 3]
```diff
@@ -134,6 +138,8 @@ def _interpolate_trace_index(ctx: InterpolationContext) -> ColorscaleInterpolant
 
 
 def _interpolate_trace_index_row_wise(ctx: InterpolationContext) -> ColorscaleInterpolants:
+    if ctx.n_rows <= 1:
```
Comment (tpvasconcelos) [inline src/ridgeplot/_color/interpolation.py:141]:
```suggestion
    if ctx.n_rows == 1:
```

**Review by tpvasconcelos — CHANGES_REQUESTED** [review]
cosmetic nit pick:

File: src/ridgeplot/_color/interpolation.py [hunk 4]
```diff
@@ -115,13 +115,17 @@ def _mul(a: tuple[Numeric, ...], b: tuple[Numeric, ...]) -> tuple[Numeric, ...]:
...
+        return [[0.0] * len(row) for row in ctx.densities]
     return [
         [((ctx.n_rows - 1) - ith_row) / (ctx.n_rows - 1)] * len(row)
         for ith_row, row in enumerate(ctx.densities)
     ]
 
 
 def _interpolate_trace_index(ctx: InterpolationContext) -> ColorscaleInterpolants:
+    if ctx.n_traces <= 1:
+        return [[0.0] * len(row) for row in ctx.densities]
```
Comment (tpvasconcelos) [inline src/ridgeplot/_color/interpolation.py:128]: Small simplification:
```suggestion
        return [[0.0]]
```

File: src/ridgeplot/_color/interpolation.py [hunk 5]
```diff
@@ -134,6 +138,8 @@ def _interpolate_trace_index(ctx: InterpolationContext) -> ColorscaleInterpolant
 
 
 def _interpolate_trace_index_row_wise(ctx: InterpolationContext) -> ColorscaleInterpolants:
+    if ctx.n_rows <= 1:
+        return [[0.0] * len(row) for row in ctx.densities]
```
Comment (tpvasconcelos) [inline src/ridgeplot/_color/interpolation.py:142]: I think here you're missing the case where the number of rows might be >1 but one of the rows only contains a single trace. Let me know if you need help with this 👍

**Comment by sstephanyy** [comment]
> @sstephanyy just noticed that you added your changes to the wrong section of the changelog. No worries of course but this would mean that your changes would not be listed in the next release notes 😨
> 
> Fixed in [bba6be2](https://github.com/tpvasconcelos/ridgeplot/commit/bba6be22c124d47c11f95fcdb1c602a01325369d) ✅

Thanks so much for the fix :)

**Comment by sstephanyy** [comment]
> @sstephanyy I simplified your test cases a bit by using `pytest.mark.parametrize` (take a look).

wow, well-done. I do know about the existence of `pytest.mark.parametrize` to run the same test over and over again with different values, but didn't come to my mind using it.  Great job :open_mouth: :confetti_ball:

File: src/ridgeplot/_color/interpolation.py (hunk 5 above)
Comment (sstephanyy) [reply src/ridgeplot/_color/interpolation.py:142]: help me with it, please 👍

File: src/ridgeplot/_color/interpolation.py (hunk 4 above)
Comment (tpvasconcelos) [reply src/ridgeplot/_color/interpolation.py]: I see you added this to all cases. Unfortunately, this simplification only applies to this specific case (`_interpolate_trace_index`). Let me revert it...

File: src/ridgeplot/_color/interpolation.py (hunk 5 above)
Comment (tpvasconcelos) [reply src/ridgeplot/_color/interpolation.py]: I'll push some test cases that should fail under this implementation. Let me know if they are helpful 👍

**Comment by tpvasconcelos** [comment]
@sstephanyy let me know if the new test cases I pushed were helpful or whether you want more help here 👍 

See [this](https://github.com/tpvasconcelos/ridgeplot/pull/268#discussion_r1838447502) as well

File: src/ridgeplot/_color/interpolation.py (hunk 5 above)
Comment (sstephanyy) [reply src/ridgeplot/_color/interpolation.py:142]: hellooo. Sorry for my late response, the last week was a busy week for me. I will start to work on it. Thank so much :)

Comment (sstephanyy) [reply src/ridgeplot/_color/interpolation.py:142]: @tpvasconcelos heellooo. I tried some solutions, but the tests still fails :/ . Sorry for that

Comment (tpvasconcelos) [reply src/ridgeplot/_color/interpolation.py]: No worries @sstephanyy! See https://github.com/tpvasconcelos/ridgeplot/pull/268/commits/31876a05a75d58cc68009e6403d9f20d15403d52 (this should fix it!)

**Review by tpvasconcelos — APPROVED** [review]
Thank you @sstephanyy for your work here!

**Pull request merged**